### PR TITLE
392 Subpixel convolutional layer

### PR DESCRIPTION
**Pull request by yiheng-wang-nv**
Implements #392 .

### Description
This implementation supports 1d, 2d and 3d inputs. The size of each dimension can be different. The implemented module is consisted with:

1. A convolutional layer, to produce a suitable number of channels for feature maps.
2. pixel shuffle, to aggregates the feature maps from low resolution space and build the super resolution space.

The pixel shuffle is modified from: [https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/PixelShuffle.cpp](url) and add 1D, 3D supports to meet MONAI's requirements.

### Status
**Ready**

### Types of changes

- [ ] Non-breaking change (fix or new feature that would not break existing functionality).
- [x] Breaking change (fix or new feature that would cause existing functionality to change).
- [x] New tests added to cover the changes.
- [x] Integration tests passed locally by running `./runtests.sh --codeformat --coverage`.
- [ ] In-line docstrings updated.
- [x] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

File: monai/networks/blocks/upsample.py [hunk 1]
```diff
@@ -71,3 +71,64 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
...
+class SubpixelUpsample(nn.Module):
+    """
+    Upsample via using a subpixel CNN. This module supports 1D, 2D and 3D input images.
+    The module is consisted with two parts. First of all, a convolutional layer is employed
+    to increase the number of channels into: ``in_channels * (scale_factor ** spatial_dims)``.
+    Secondly, a pixel shuffle manipulation is utilized to aggregates the feature maps from
+    low resolution space and build the super resolution space.
+    The first part of the module is not fixed, a sequential layers can be used to replace a
+    single layer. 
+    The idea comes from:
```
Comment (wyli) [inline monai/networks/blocks/upsample.py:85]: not fixing the conv part will provide more flexibility, but seems it's not implemented yet in this PR?

File: monai/networks/blocks/upsample.py [hunk 2]
```diff
@@ -71,3 +71,64 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
...
+        return x
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        """
+        Args:
+            x: Tensor in shape (batch, channel, spatial_1[, spatial_2, ...).
+        """
+        x = self.conv(x)
+        x = self.pixelshuffle(x)
+        return torch.as_tensor(x)
```
Comment (wyli) [inline monai/networks/blocks/upsample.py:134]: just for clarification, any particular reason to use `torch.as_tensor` here?

File: monai/networks/blocks/upsample.py [hunk 3]
```diff
@@ -71,3 +71,64 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
...
+        """
+        Args:
+            spatial_dims: number of spatial dimensions of the input image.
+            in_channels: number of channels of the input image.
+            scale_factor: multiplier for spatial size. Defaults to 2.
+        """
+        super().__init__()
+        conv_out_channels = in_channels * (scale_factor ** spatial_dims)
+        self.spatial_dims = spatial_dims
+        self.scale_factor = scale_factor
```
Comment (wyli) [inline monai/networks/blocks/upsample.py:103]: need some validation such as `scale_factor > 1`?

File: monai/networks/blocks/upsample.py (hunk 1 above)
Comment (yiheng-wang-nv) [reply monai/networks/blocks/upsample.py:85]: yes, not implement. I'm not sure how to write down a flexible structure here, cuz users can use 1 or 2 or many different layers before the pixel shuffle part.

File: monai/networks/blocks/upsample.py (hunk 2 above)
Comment (yiheng-wang-nv) [reply monai/networks/blocks/upsample.py:134]: Nope, I just follow the representation in ``Upsample`` class. Let me remove it if needed.

File: monai/networks/blocks/upsample.py (hunk 3 above)
Comment (yiheng-wang-nv) [reply monai/networks/blocks/upsample.py:103]: That makes sense, let me add it.
